expose gradual type inferface to HFtracer and generate constraints for gt nodes

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81360
* #81344
* #81274
* #81265
* #81190
* #81189
* #81175
* #81159
* #80976
* #80925
* #80909
* #80847
* #80823
* #80799
* __->__ #80744

Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets.

migeed-z committed:
expose gradual type inferface to HFtracer and generate constraints for gt nodes

[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"




[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"




[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

jamesr66a commented on torch/fx/experimental/migrate_gradual_types/transform_to_z3.py:
```diff
@@ -240,5 +240,84 @@ def transform_all_constraints(traced, counter=0):
+        old_c = None
+        while old_c != condition_constraint:
+            old_c = condition_constraint
+            condition_constraint, counter = transform_constraint(condition_constraint, counter)
```
As discussed offline, we probably want to hoist the fixpoint portion of this into a function so we don't a) have it repeated and b) expose low-level details

migeed-z replied:
done

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z committed:
Update on "expose gradual type inferface to HFtracer and generate constraints for gt nodes"


Define the function `transform_all_constraints_trace_time` which takes a node which represents a conditional, a graph and a root which represent the trace so far. We then generate constraints for the current trace, and a separate constraint for the conditional. We then combine the constraints with the node constraint in one set and we combine the constraint with the negation of the node's constraint in another set. We then transform both sets to Z3 format and we return both sets. 


[ghstack-poisoned]

migeed-z commented:
@pytorchbot merge -g
